examples/piecestore/rpc/client/main.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/urfave/cli"
+	"github.com/zeebo/errs"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc"
+
+	"storj.io/storj/examples/piecestore/rpc/client/utils"
+	"storj.io/storj/pkg/piecestore/rpc/client"
+	pb "storj.io/storj/protos/piecestore"
+)
+
+var argError = errs.Class("argError")
+
+func main() {
+
+	app := cli.NewApp()
+
+	// Set up connection with rpc server
+	var conn *grpc.ClientConn
+	conn, err := grpc.Dial(":7777", grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("did not connect: %s", err)
+	}
+	defer conn.Close()
+	routeClient := client.New(context.Background(), pb.NewPieceStoreRoutesClient(conn))
+
+	app.Commands = []cli.Command{
+		{
+			Name:    "upload",
+			Aliases: []string{"u"},
+			Usage:   "upload data",
+			Action: func(c *cli.Context) error {
+				if c.Args().Get(0) == "" {
+					return argError.New("No input file specified")
+				}
+
+				file, err := os.Open(c.Args().Get(0))
+				if err != nil {
+					return err
+				}
+				// Close the file when we are done
+				defer file.Close()
+
+				fileInfo, err := file.Stat()
+				if err != nil {
+					return err
+				}
+
+				if fileInfo.IsDir() {
+					return argError.New(fmt.Sprintf("Path (%s) is a directory, not a file", c.Args().Get(0)))
+				}
+
+				var fileOffset, storeOffset int64 = 0, 0
+				var length = fileInfo.Size()
+				var ttl = time.Now().Unix() + 86400
+
+				id, err := utils.DetermineID(file, fileOffset, length)
+				if err != nil {
+					return err
+				}
+
+				// Created a section reader so that we can concurrently retrieve the same file.
+				dataSection := io.NewSectionReader(file, fileOffset, length)
+
+				writer, err := routeClient.StorePieceRequest(id, fileOffset, length, ttl, storeOffset)
+				if err != nil {
+					fmt.Printf("Failed to send meta data to server to store file of id: %s\n", id)
+					return err
+				}
+
+				_, err = io.Copy(writer, dataSection)
+				if err != nil {
+					fmt.Printf("Failed to store file of id: %s\n", id)
+				} else {
+					fmt.Printf("successfully stored file of id: %s\n", id)
+				}
+
+				return writer.Close()
+			},
+		},
+		{
+			Name:    "download",
+			Aliases: []string{"d"},
+			Usage:   "download data",
+			Action: func(c *cli.Context) error {
+				if c.Args().Get(0) == "" {
+					return argError.New("No id specified")
+				}
+
+				id := c.Args().Get(0)
+
+				if c.Args().Get(1) == "" {
+					return argError.New("No output file specified")
+				}
+
+				_, err := os.Stat(c.Args().Get(1))
+				if err != nil && !os.IsNotExist(err) {
```

Comment:
what if err == nil? do we want to return some sort of argError that the destination exists?

Reply:
not important enough to hold up the merge

Reply:
👍